pkg/controller/memberstatus/memberstatus_controller.go

```diff
@@ -0,0 +1,236 @@
+	c, err := controller.New("memberstatus-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// watch for changes to primary resource MemberStatus
+	err = c.Watch(&source.Kind{Type: &toolchainv1alpha1.MemberStatus{}}, &handler.EnqueueRequestForObject{}, predicate.GenerationChangedPredicate{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileMemberStatus implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileMemberStatus{}
+
+// ReconcileMemberStatus reconciles a MemberStatus object
+type ReconcileMemberStatus struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client         client.Client
+	scheme         *runtime.Scheme
+	getHostCluster func() (*cluster.FedCluster, bool)
+}
+
+// Reconcile reads the state of toolchain member cluster components and updates the MemberStatus resource with information useful for observation or troubleshooting
+func (r *ReconcileMemberStatus) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling MemberStatus")
+
+	// fetch the MemberStatus
+	memberStatus := &toolchainv1alpha1.MemberStatus{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, memberStatus)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// MemberStatus resource not found, log error and return
+			reqLogger.Error(err, "MemberStatus resource was not found")
+		}
+		return reconcile.Result{}, nil
+	}
+
+	err = r.aggregateAndUpdateStatus(reqLogger, memberStatus)
+	if err != nil {
+		reqLogger.Error(err, "Failed to update status")
+	}
+
+	reqLogger.Info(fmt.Sprintf("Finished updating MemberStatus, requeueing after %v", defaultRequeueTime))
+	return reconcile.Result{RequeueAfter: defaultRequeueTime}, nil
+}
+
+type statusHandler struct {
+	Name         statusComponentTag
+	handleStatus func(logger logr.Logger, memberStatus *toolchainv1alpha1.MemberStatus) error
+}
+
+func (r *ReconcileMemberStatus) aggregateAndUpdateStatus(reqLogger logr.Logger, memberStatus *toolchainv1alpha1.MemberStatus) error {
+
+	memberOperatorStatusHandler := statusHandler{Name: memberOperator, handleStatus: r.memberOperatorHandleStatus}
+	hostConnectionStatusHandler := statusHandler{Name: hostConnection, handleStatus: r.hostConnectionHandleStatus}
+
+	statusHandlers := []statusHandler{memberOperatorStatusHandler, hostConnectionStatusHandler}
+
+	// track components that are not ready
+	unreadyComponents := []string{}
+
+	// retrieve component statuses eg. kubefed, member deployment
+	for _, handler := range statusHandlers {
+		err := handler.handleStatus(reqLogger, memberStatus)
+		if err != nil {
+			reqLogger.Error(err, "status update problem")
+			unreadyComponents = append(unreadyComponents, string(handler.Name))
```

Comment:
The errors returned by the handlers seem to get lost (they are only logged). I don't see the error messages to be added to the MemberStatus.

It seems that there is a mix of different kind of errors in both handlers: 
errors which should go to the status when one of the component is not ready from an error when we can't really obtain the status (there is a problem with the client for example or there is no KubeFed cluster resource at all).

I would rather add any error we get when trying to update the component status to the corresponding component status. In this case a handler don't really need to return any error (all errors go to the component) but return a bool instead (component ready, or not). If any handler returns `false` then the overall condition is not ready.

Reply:
Updated it so that now all the errors are visible in the component statuses. I kept the status handlers returning errors instead of bool because we're currently logging the error which I think is still useful in case some errors show up briefly and disappear? But I can change it if we'd rather handle it differently.

So with the change when the kubefed resource or deployment can't be found the error is now available in the component status:
```
status:
  conditions:
  - lastTransitionTime: "2020-06-24T20:37:47Z"
    lastUpdatedTime: "2020-06-24T20:37:47Z"
    message: 'Components not ready: [memberOperator hostConnection]'
    reason: ComponentsNotReady
    status: "False"
    type: Ready
  hostConnection:
    conditions:
    - lastProbeTime: "2020-06-24T20:37:47Z"
      message: the host connection was not found
      reason: KubefedNotFound
      status: "False"
      type: Ready
  memberOperator:
    buildTimestamp: "2020-06-24T20:35:11Z"
    conditions:
    - lastTransitionTime: "2020-06-24T20:37:47Z"
      lastUpdatedTime: "2020-06-24T20:37:47Z"
      message: unable to get the member operator deployment
      reason: DeploymentNotFound
      status: "False"
      type: Ready
    deployment:
      name: ""
    revision: f617c6207d45e7f528f250ccae0ac2fe72fa17d6-dirty
    version: 0.0.1
```